actors/builtin/market/market_test.go

```diff
@@ -352,9 +354,289 @@ func TestPublishStorageDeals(t *testing.T) {
+			signatureVerificationError error
+		}{
+			"deal end after deal start": {
+				setup: func(_ *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.StartEpoch = 10
+					d.EndEpoch = 9
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"current epoch greater than start epoch": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					rt.SetEpoch(d.StartEpoch + 1)
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"deal duration greater than max deal duration": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.StartEpoch = abi.ChainEpoch(10)
+					d.EndEpoch = d.StartEpoch + (1 * builtin.EpochsInYear) + 1
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"negative price per epoch": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.StoragePricePerEpoch = abi.NewTokenAmount(-1)
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"price per epoch greater than total filecoin": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.StoragePricePerEpoch = big.Add(abi.TotalFilecoin, big.NewInt(1))
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"negative provider collateral": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.ProviderCollateral = big.NewInt(-1)
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"provider collateral greater than max collateral": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.ProviderCollateral = big.Add(abi.TotalFilecoin, big.NewInt(1))
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"negative client collateral": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.ClientCollateral = big.NewInt(-1)
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"client collateral greater than max collateral": {
+				setup: func(rt *mock.Runtime, _ *marketActorTestHarness, d *market.DealProposal) {
+					d.ClientCollateral = big.Add(abi.TotalFilecoin, big.NewInt(1))
+				},
+				exitCode: exitcode.ErrIllegalArgument,
+			},
+			"client does not have enough balance for collateral": {
+				setup: func(rt *mock.Runtime, a *marketActorTestHarness, d *market.DealProposal) {
+					a.addParticipantFunds(rt, client, big.Sub(d.ClientBalanceRequirement(), big.NewInt(1)))
+					a.addProviderFunds(rt, d.ProviderCollateral)
+				},
+				exitCode: exitcode.ErrInsufficientFunds,
+			},
+			"provider does not have enough balance for collateral": {
+				setup: func(rt *mock.Runtime, a *marketActorTestHarness, d *market.DealProposal) {
+					a.addParticipantFunds(rt, client, d.ClientBalanceRequirement())
+					a.addProviderFunds(rt, big.Sub(d.ProviderCollateral, big.NewInt(1)))
```

Comment:
This is ok, but also consider doing this the other way too: set up the provider/participant with some funds, fixed for all cases, then change the deal to cost just too much.

Reply:
This is done.